utils/utf8.cc

```diff
@@ -56,22 +56,30 @@ namespace utils {
 
 namespace utf8 {
 
-// 3x faster than boost utf_to_utf
-static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t len) {
-    size_t pos = 0;
-
-    while (len) {
-        size_t bytes;
-        const uint8_t byte1 = data[0];
+struct codepoint_status {
+    size_t bytes_validated;
+    bool error;
+    uint8_t more_bytes_needed;
+};
 
+static
+codepoint_status
+inline
+evaluate_codepoint(const uint8_t* data, size_t len) {
+    auto byte1 = data[0];
+    static uint8_t len_from_first_nibble[16] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
+    auto codepoint_len = len_from_first_nibble[byte1 >> 4];
+    if (codepoint_len > len) {
+	return codepoint_status{.more_bytes_needed = uint8_t(codepoint_len - len)};
+    } else {
         if (byte1 <= 0x7F) {
             // 00..7F
-            bytes = 1;
-        } else if (len >= 2 && byte1 >= 0xC2 && byte1 <= 0xDF &&
+            return codepoint_status{.bytes_validated = codepoint_len};
+        } else if (byte1 >= 0xC2 && byte1 <= 0xDF &&
                 (int8_t)data[1] <= (int8_t)0xBF) {
             // C2..DF, 80..BF
-            bytes = 2;
-        } else if (len >= 3) {
+            return codepoint_status{.bytes_validated = codepoint_len};
+        } else {
             const uint8_t byte2 = data[1];
 
             // Is byte2, byte3 between 0x80 ~ 0xBF
```

Comment:
This comment says that we check bytes are within [0x80, 0xBF] but the check below compares only against 0xBF. Shouldn't we also compare against 0x80?

Reply:
All int8_t values are >= 0x80, that is the minimum value.

Reply:
It's not that all int8_t values are >= 0x80. It's more that every value that's < 0x80 in uint8_t is bigger than 0xBF in int8_t. Otherwise, the check wouldn't work.

But you're right everything's fine.

Reply:
godbolt thinks it is:

   https://godbolt.org/z/9rKah7

0xbf checks the other side of the interval. The trick reduced a closed interval into a half-open interval.

Reply:
Compiler optimization is understandable. 0x80 is the minimal value of int8_t so there's no point checking that something is smaller than 0x80. But it's all about the conversion from uint8_t to int8_t. Values that are < 0x80 in uint8_t after conversion to int8_t become > than 0xBF and that's why this trick works. If values that are < 0x80 in uint8_t after conversion to int8_t were taking values smaller than 0xBF then the trick wouldn't work. The `<= 0xBF` check serves two purposes:
1. eliminate all values that in uint8_t are bigger than 0xBF because they are still bigger than 0xBF after conversion to int8_t
2. eliminate all values that in uint8_t are smaller than 0x80 because after conversion to int8_t they become bigger than 0xBF

What I'm saying is that the fact that 0x80 is a minimal value for int8_t does not make the trick correct. What does is the way uint8_t is converted to int8_t. It's based on the fact that the conversion turns every x < 0x80 into y > 0xBF. If it wasn't true the trick wouldn't work even with 0x80 being the minimal value of int8_t.

Reply:
```
const int byte2_ok = (int8_t)byte2 <= (int8_t)0xBF;
```
Wow, what a trick... Basically, 0xBF is a negative number, so for byte2 to be smaller than it, it needs to be negative too (i.e., the unsigned version is >= 0x80), and must also be <= 0xBF. This should have been in a separate function with clear comments explaining how/why it works. I agree it's pre-existing code, so you don't *have* to fix it too, but maybe it's a good opportunity since we understand it now...

Reply:
I agree it's very hacky. Hard to spot.

Reply:
The same trick is used in the SIMD code, where it's not possible to extract.